commands/src/main/java/org/wildfly/extras/creaper/commands/foundation/online/SnapshotBackup.java

```diff
@@ -0,0 +1,125 @@
+package org.wildfly.extras.creaper.commands.foundation.online;
+
+import java.io.IOException;
+import java.util.concurrent.TimeoutException;
+
+import org.jboss.logging.Logger;
+import org.wildfly.extras.creaper.core.CommandFailedException;
+import org.wildfly.extras.creaper.core.online.CliException;
+import org.wildfly.extras.creaper.core.online.ModelNodeResult;
+import org.wildfly.extras.creaper.core.online.OnlineCommand;
+import org.wildfly.extras.creaper.core.online.OnlineCommandContext;
+import org.wildfly.extras.creaper.core.online.operations.admin.ReloadToSnapshot;
+
+import com.google.common.base.Strings;
+
+/**
+ * Provides a pair of online commands to backup and then restore application
+ * server configuration while the server is running. Backup consists of taking a
+ * snapshot, restore then means reloading the server from the snapshot.
+ *
+ * <p>
+ * The {@code backup} command must be applied before {@code restore},
+ * {@code backup} can be applied once, and then {@code restore} can be applied
+ * many times. If any one of these rules is violated, an exception is thrown.
+ * For special circumstances, when the backup that was already acquired is no
+ * longer needed and is not going to be restored, a {@code destroy} command is
+ * provided. If there was no backup acquired, the destroy command does nothing.
+ * </p>
+ */
+public final class SnapshotBackup {
+    private static final Logger log = Logger.getLogger(SnapshotBackup.class);
+
+    private String snapshotName; // null <=> backup wasn't acquired, can't restore
+
+    private final OnlineCommand backupPart = new OnlineCommand() {
+        @Override
+        public void apply(OnlineCommandContext occ) throws CommandFailedException, IOException, CliException {
+            if (SnapshotBackup.this.snapshotName != null) {
+                throw new CommandFailedException("Snapshot was already taken: "
+                        + SnapshotBackup.this.snapshotName);
+            }
+            String snapshotAbsolutePath = "";
+            ModelNodeResult modelNodeResult = occ.client.execute(":take-snapshot");
+            if (modelNodeResult.isSuccess() && modelNodeResult.hasDefinedValue()) {
+                snapshotAbsolutePath = modelNodeResult.stringValue();
+            }
+            // returns absolute path, but we need just the name of snapshot
+            String snapshotName = getSnapshotName(snapshotAbsolutePath);
+            if (Strings.isNullOrEmpty(snapshotAbsolutePath)) {
+                throw new CommandFailedException("Unable to take snapshot");
```

Comment:
`OnlineManagementClient.execute` won't throw in case of an error. However, you can do `ModelNodeResult.assertSuccess()`, which would be enough in this case IMHO.